src/Stratis.Bitcoin/BlockPulling2/BlockPullerBehavior.cs

```diff
@@ -0,0 +1,245 @@
+            int penalizeTimes = notDeliveredBlocksCount < maxSamplesToPenalize ? notDeliveredBlocksCount : maxSamplesToPenalize;
+            
+            this.logger.LogDebug("Peer will be penalized {0} times.", penalizeTimes);
+
+            for (int i = 0; i < penalizeTimes; i++)
+                this.AddSample(0, delaySeconds);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        public void OnIbdStateChanged(bool isIbd)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(isIbd), isIbd);
+
+            // Recalculates the max samples count that can be used for quality score calculation.
+            int samplesCount = isIbd ? IbdSamplesCount : NormalSamplesCount;
+            this.averageSizeBytes.SetMaxSamples(samplesCount);
+            this.averageDelaySeconds.SetMaxSamples(samplesCount);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        public void RecalculateQualityScore(int bestSpeedBytesPerSecond)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(bestSpeedBytesPerSecond), bestSpeedBytesPerSecond);
+
+            this.QualityScore = (double)this.SpeedBytesPerSecond / bestSpeedBytesPerSecond;
+
+            if (this.QualityScore < MinQualityScore)
+                this.QualityScore = MinQualityScore;
+
+            if (this.QualityScore > MaxQualityScore)
+                this.QualityScore = MaxQualityScore;
+
+            this.logger.LogTrace("Quality score was set to {0}.", this.QualityScore);
+            this.logger.LogTrace("(-)");
+        }
+        
+        private Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(message), message.Message.Command);
+
+            if (message.Message.Payload is BlockPayload block)
+            {
+                uint256 blockHash = block.Obj.GetHash();
+
+                this.logger.LogTrace("Block '{0}' delivered.", blockHash);
+                this.blockPuller.PushBlock(blockHash, block.Obj, peer.Connection.Id);
+            }
+
+            this.logger.LogTrace("(-)");
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc/>
+        public async Task RequestBlocksAsync(List<uint256> hashes)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(hashes), nameof(hashes.Count), hashes.Count);
+
+            var getDataPayload = new GetDataPayload();
+
+            foreach (uint256 uint256 in hashes)
+            {
+                var vector = new InventoryVector(InventoryType.MSG_BLOCK, uint256);
+                vector.Type = this.AttachedPeer.AddSupportedOptions(vector.Type);
+
+                getDataPayload.Inventory.Add(vector);
+            }
+
+            if (this.AttachedPeer.State != NetworkPeerState.HandShaked)
```

Comment:
Is this check really needed?
Where do you catch this?and I guess it means the caller needs to release the hashes to another peer.

Reply:
It is caught at BlockPuller:570
In that case assignments will be released.

Checking for handshaked state is needed because otherwise we know that message won't be processed by other peer according to the protocol. This check was present in the prev block puller behavior.